### ISSUE TITLE
Obfuscate signs based on whether the sign has text or not.

### ISSUE THREAD

**Issue by MathInTheBath** [lishid/Orebfuscator#146]
Signs are used both for giving directions and for filling space desired to obfuscated. The issue occurs where signs are used for underground navigation, but players far away can't tell that they're signs due to obfuscation, making navigation cumbersome and ugly. This is an issue on Devoted because signs are obfuscated below y=94. The compromise is this: If a sign has no text, obfuscate it normally, and otherwise, don't obfuscate it.

**Comment by ProgrammerDan** [comment]
s/fence/sign/gc

**Comment by MathInTheBath** [comment]
@ProgrammerDan I could lie and say English isn't my first language... but... let's just say I'm tired.

**Comment by ProgrammerDan** [comment]
@Aleksey-Terzi one thought brought up in discussion (no idea if feasible) would be if possible to set a "treat-empty-sign-as" block and a "treat-filled-sign-as" block, so you could have independent obfuscation behavior based on the detected contents of the sign, without having to modify any of the check-code to determine obfuscation later.

**Comment by Aleksey-Terzi** [comment]
Unfortunately "empty sign" and "sign with text" treated as the same block.
Also looks like "meta" haven't information about this as well (if we would supported "meta" for obfuscation, but we do not support it)

Look like the only way is to read data form sign entity.
I don't remember exactly if we are receiving entity information in the same packet or no, but I'm sure this feature is doable.
The other question what impact will have this feature to performance and how to make it so this feature is universal for all entity-based blocks.

Also I'm not sure if this will be reasonable to have such feature.
Idea was to obfuscate signs in the warehouses.
But if this feature is not demanded (I don't have signs in my warehouses at all :) ), then sign block (two block types) could be removed from obfuscation in config file.
Seems the only cases where empty sign could be used are traps and water entrances.

**Comment by ProgrammerDan** [comment]
BlockState for Signs includes sign text information, example: https://github.com/DevotedMC/Humbug/blob/master/src/main/java/com/untamedears/humbug/Humbug.java#L2675

Sadly we can't remove signs from config. People use them to disguise underground bunkers (to hide the aie pocket).

The above example from humbug is why I recommended if it was possible to have a replacement check during the initial readthrough of BlockState -- if sign with text, treat like BANNER; if sign without text, treat like SIGN/CHEST, etc.

But I have a very uneducated layman's understanding of what you're doing here in Obfuscator, so if it's not possible, then I concede the point :D

It's definitely not worth an enormous restructure to support.

**Comment by Aleksey-Terzi** [comment]
> BlockState for Signs includes sign text information, example: https://github.com/DevotedMC/Humbug/blob/master/src/main/java/com/untamedears/humbug/Humbug.java#L2675

Yes, but this information anyway is going from TileEntity:

![image](https://cloud.githubusercontent.com/assets/5383652/24048832/c31b605e-0b32-11e7-8f52-c483484107de.png)

> But I have a very uneducated layman's understanding of what you're doing here in Obfuscator, so if it's not possible, then I concede the point :D

Orebfuscator is catching this chunk data packet when it is sending to client: http://wiki.vg/Protocol#Chunk_Data
and then processing it here: https://github.com/lishid/Orebfuscator/blob/master/Plugin/src/main/java/com/lishid/orebfuscator/obfuscation/Calculations.java#L40

For now Orebfuscator is doing manipulations just with "Data" field (format is here: http://wiki.vg/SMP_Map_Format)

But probably we can also use information from "Block entities" field (yesss, it is in the same packet) to determine if sign has text or no.

But how this should work from administrator's view?
How this should be told in config or etc?
In other words how you (as administrator) want to tell to Orebfuscator that it should skip signs with text without obfuscation?

**Comment by ProgrammerDan** [comment]
Very cool, thank you for the walkthrough.

From an administrator's view, I have two conflicting opinions:

1) Just have a boolean `BypassObfuscationForSignsWithText: true/false` -- set that per world like other booleans

OR

2) Have a "obfuscate-as" -- `ObfuscateSignsWithTextAs: BLOCK_TYPE` -- set that per world, and in the obfuscation engine, if the sign has text, pretend it's the BLOCK_TYPE set in the config instead during computation of obfuscation.

I think for me either would be fine, I'm assuming the boolean would be simpler if this is possible at all.

What's the complexity impact? Would this increase running time dramatically for each ChunkData processed?

**Comment by ProgrammerDan** [comment]
Any new thoughts on this @Aleksey-Terzi ?

**Comment by Aleksey-Terzi** [comment]
Sorry, I will try to reply either later today or tomorrow.

**Comment by Aleksey-Terzi** [comment]
> I think for me either would be fine, I'm assuming the boolean would be simpler if this is possible at all.

Yes, lets try first variant.
Also I'm not fully understanding meaning of the second variant with BLOCK_TYPE :)

> What's the complexity impact? Would this increase running time dramatically for each ChunkData processed?

I hopes if everything will be done right then it will not reflect performance (at least significantly).
We have data block and just need to extract some info from it, check it and then according to result decide if sign block (actually two block types used for sign) should be obfuscated or no (if sign is in list of blocks to obfuscate).
Seems everything is simple, but need to spend some time for experiments.
I will try to implement this feature ASAP, probably these weekends but can't promise.

**Comment by Aleksey-Terzi** [comment]
Done, pushed to Devoted fork.

By default it works as previously (i.e. BypassObfuscationForSignsWithText = false), so you should set **true** value to apply new functionality.

Function works as I supposed previously - obfuscation module has all block NBT tags inside the packet, so it is not doing any outside requests and obfuscation should work with the same performance as previously.

@ProgrammerDan I suggest to make new release and update topic on spigot website when you will have the time.
If you have any questions please ask, if not then close issue.

**Comment by ProgrammerDan** [comment]
@Aleksey-Terzi first stage in release process complete: https://github.com/lishid/Orebfuscator/releases/tag/4.2.2

Please tag me in 2 days if spigot and bucket releases haven't dropped by then.
